aesd-char-driver/main.c

```diff
@@ -0,0 +1,339 @@
+        struct aesd_dev *dev = filp->private_data;
+        ssize_t offs_in_found = 0; //will be the offset in the found command that fpos points to 
+        struct aesd_buffer_entry *found_entry;
+        ssize_t bytes_to_read = 0;
+
+        PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+        /**
+         * TODO: handle read
+         */
+
+        //Reference: scull main.c
+        //return if mutex wait interrupted
+        if (mutex_lock_interruptible(&dev->lock))
+            return -ERESTARTSYS;
+
+        //Find the entry and offset within that entry corresponding to f_pos
+
+        found_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&(dev->circ_buff), *f_pos, &offs_in_found);
+
+        if (found_entry == NULL){
+            retval = 0; 
+            //we don't have seek capability yet to move to the start of the buffer after a read is finished and we are doing another.
+            //So assume that if the reader has gotten here, they have already read the entire file and are trying one more read checking if bytes returned was 0. 
+            //reset f_pos according to aesd_socket application's necessary operation. 
+            //*f_pos = 0; //When we have seek implemented, or if our char driver had other usecases we could remove this. 
+            goto read_end;
+        }
+
+        //Read from fpos to end of entry and update fpos
+        bytes_to_read = ((found_entry->size) - offs_in_found);
+        if (count < bytes_to_read)
+            bytes_to_read = count;
+
+        if ( copy_to_user(buf, ((found_entry->buffptr) + offs_in_found), bytes_to_read) ){
+            retval = -EFAULT;
+            goto read_end;
+        }
+
+        *f_pos +=  bytes_to_read;
+        retval = bytes_to_read;
+
+    read_end:
+        //unlock lock here...
+        mutex_unlock(&dev->lock);
+        //PDEBUG("read returning with %zu bytes read", retval);
+        //PDEBUG("filepos after read: %lld",*f_pos);
+        return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+        ssize_t retval = -ENOMEM;
+        struct aesd_dev *dev = filp->private_data;
+        char * new_write;
+        const char * old_buffer;
+        char * full_buffer; //will hold old_buffer + new_write when \n received.
+        char * newl_ptr; //Will be a pointer to the newline char in a new write if found
+        ssize_t len_cmd; //Length up to and including the newline if a newline recvd
+        ssize_t full_buffer_size; //size of full buffer when we are about to write it to circular buffer
+        ssize_t i;
+
+        PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+        /**
+         * TODO: handle write
+         */
+
+        //Reference: scull main.c
+        //return if mutex wait interrupted
+        if (mutex_lock_interruptible(&dev->lock))
+            return -ERESTARTSYS;
+
+        new_write = (char *)kmalloc(count, GFP_KERNEL);
```

Comment:
@malcolmmckellips does this block and the one below it need to be inside the mutex lock?  What is dev->lock locking?